samples: Add TF-M sample for custom secure partitions

oyvindronningstad opened the pull request:
The sample implements a dummy ARoT (Application Root of Trust) partition in
TF-M and calls it from the app.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

oyvindronningstad committed:
samples: Add TF-M sample for custom secure partitions

The sample implements a dummy ARoT (Application Root of Trust) partition
in TF-M and calls it from the app.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

microbuilder commented:
@oyvindronningstad I didn't realise you were working on this problem as well, and @urutva put together a PR for the same thing: https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/55

MirkoCovizzi commented on samples/tfm_integration/tfm_secure_partition/dummy_partition/dummy_partition.c:
```diff
@@ -0,0 +1,205 @@
+		return PSA_ERROR_PROGRAMMER_ERROR;
+	}
+
+	callback(handle, digest, digest_size);
```
This callback procedure returns nothing, you should return PSA_SUCCESS afterwards.

joerchan replied:
Thanks. fixed.

hakonfam commented on samples/tfm_integration/tfm_secure_partition/CMakeLists.txt:
```diff
@@ -0,0 +1,30 @@
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+
+configure_file(
+  ${CMAKE_CURRENT_LIST_DIR}/dummy_partition/tfm_manifest_list.yaml
```
Is there any convention with how to name a file that is configured in this way? IMO it would be good for them to have different names so that it is easy to see which is which when grepping or searching for files.

joerchan replied:
Common to append `.in` to the name. Renamed.

hakonfam commented on samples/tfm_integration/tfm_secure_partition/README.rst:
```diff
@@ -0,0 +1,50 @@
+This sample creates a dummy secure partition and secure service for TF-M and instructs the TF-M build system to build it into the secure firmware.
+The dummy secure service is then called in the main file (in the non-secure firmware).
+
+Everything in the "partition" directory is built by the TF-M build system (behind the scenes), for the secure firmware image.
```
Should we point to the module integration and mention that its built as an external project instead of `behind the scenes` ?

hakonfam commented on samples/tfm_integration/tfm_secure_partition/dummy_partition/CMakeLists.txt:
```diff
@@ -0,0 +1,50 @@
+
```
Missing copyright?

SebastianBoe reviewed (changes requested):
This sample and corresponding patch to upstream TFM prompted upstream TFM to add a slightly different support for out-of-tree secure partitions.

Great work @oyvindronningstad .

This PR should be re-written to use this new mechanism.

https://review.trustedfirmware.org/c/TF-M/trusted-firmware-m/+/10562

urutva commented:
@SebastianBoe I had a discussion with @microbuilder regarding bringing in cmake out-of-tree build changes that didn't make it to 1.4.0/1 release. @microbuilder said, we can either cherry-pick the [commit](https://git.trustedfirmware.org/TF-M/trusted-firmware-m.git/commit/?id=b269420edcafe6c87454e833776cbc598227e41f) or do an upmerge till that commit.

Since 1.4.1 upmerge `[PR](https://github.com/zephyrproject-rtos/trusted-firmware-m/pull/61)` is merged now, should we start looking into brining in cmake out-of-tree build changes to un-block this PR?

SebastianBoe commented:
Hi, we are actively working on this now, but don't have a strategy yet.

A concern with upmerging to a commit is that security advisories specify the release versions that are affected and fixed. So using a commit can make it difficult to understand if one is affected.

cherry-picking build system changes sounds a lot safer, albeit with higher risk of dealing with conflicts.

Will get back to you when we know what we are going to do.

SebastianBoe commented on samples/tfm_integration/tfm_secure_partition/README.rst:
```diff
@@ -0,0 +1,49 @@
+
+This dummy partition has a single secure service, which can index one of 5 dummy secrets inside the partition, and retrieve a hash of the secret.
+
+When adapting this partition for your own purposes, please change all occurences of "dummy_partition", "DUMMY_PARTITION", "dp", and "DP" to your own partition name. Also, look through the CMakeLists.txt both for the secure and non-secure and make relevant changes, as well as the yaml files inside "partition".
```
look for the secure what?

joerchan replied:
"Also, look through the CMakeLists.txt both for the secure and non-secure and make relevant changes"
Would be better to invert this to make it easier to read.
"Also, look through both the secure and non-secure CMakeLists.txt file and make relevant changes"

SebastianBoe commented on samples/tfm_integration/tfm_secure_partition/README.rst:
```diff
@@ -0,0 +1,49 @@
+This sample creates a dummy secure partition and secure service for TF-M and instructs the TF-M build system to build it into the secure firmware.
+The dummy secure service is then called in the main file (in the non-secure firmware).
+
+Everything in the "partition" directory is built by the TF-M build system, refer to :ref:`tfm_build_system` for more details.
```
I thought this was referring to a directory literally named "partition". Can we make this clearer?

joerchan replied:
I've restructured the text, hopefully this is clearer.

SebastianBoe commented on samples/tfm_integration/tfm_secure_partition/dummy_partition/CMakeLists.txt:
```diff
@@ -0,0 +1,54 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+add_library(tfm_app_rot_partition_dp STATIC)
```
Cleaner to omit the target_sources invocations below and place the sources straight into the add_library invocation.

joerchan replied:
This follows TF-Ms general pattern for adding partition sources. I prefer to stick to it for consistency for that reason.

SebastianBoe commented on samples/tfm_integration/tfm_secure_partition/dummy_partition/tfm_manifest_list.yaml.in:
```diff
@@ -0,0 +1,30 @@
+    {
+      "name": "Dummy Partition",
+      "short_name": "TFM_DP",
+      "manifest": "${ZEPHYR_BASE}/samples/tfm_integration/tfm_secure_partition/dummy_partition/tfm_dummy_partition.yaml",
```
```suggestion
      "manifest": "${APPLICATION_SOURCE_DIR}/dummy_partition/tfm_dummy_partition.yaml",
```

SebastianBoe commented on samples/tfm_integration/tfm_secure_partition/dummy_partition/tfm_manifest_list.yaml.in:
```diff
@@ -0,0 +1,30 @@
+      "name": "Dummy Partition",
+      "short_name": "TFM_DP",
+      "manifest": "${ZEPHYR_BASE}/samples/tfm_integration/tfm_secure_partition/dummy_partition/tfm_dummy_partition.yaml",
+      "source_path": "${ZEPHYR_BASE}/samples/tfm_integration",
```
weird. But if they want this then:

${APPLICATION_SOURCE_DIR}/..

SebastianBoe replied:
Might want to double-check that this is correct as it is strange.

SebastianBoe commented on samples/tfm_integration/tfm_secure_partition/dummy_partition/tfm_manifest_list.yaml.in:
```diff
@@ -0,0 +1,30 @@
+      "short_name": "TFM_DP",
+      "manifest": "${ZEPHYR_BASE}/samples/tfm_integration/tfm_secure_partition/dummy_partition/tfm_dummy_partition.yaml",
+      "source_path": "${ZEPHYR_BASE}/samples/tfm_integration",
+      "tfm_partition_ipc": true,
```
Generate this from a Kconfig perhaps?

SebastianBoe replied:
Never mind, I misunderstood. Looks good.

- ``tfm_partition_ipc``: indicate if this partition is compatible with the IPC
  model.

SebastianBoe commented on samples/tfm_integration/tfm_secure_partition/dummy_partition/tfm_manifest_list.yaml.in:
```diff
@@ -0,0 +1,30 @@
+        "library_list": [
+           "*tfm_*partition_dp.*"
+         ],
+      }
```
is pid not required?

SebastianBoe commented on samples/tfm_integration/tfm_secure_partition/dummy_partition/dummy_partition.c:
```diff
@@ -0,0 +1,212 @@
+static psa_status_t tfm_dp_init(void)
+{
+	return TFM_SUCCESS;
+}
```
This is weird, can it be deleted?

joerchan replied:
Yes, this can be deleted.

SebastianBoe commented on samples/tfm_integration/tfm_secure_partition/dummy_partition/dummy_partition.c:
```diff
@@ -0,0 +1,212 @@
+	}
+
+	return PSA_SUCCESS;
+}
```
Can we remove this also?

Either I am reading the code wrong, or at some point this was necessary and then it got reworked to not be needed.

joerchan committed:
cmake: Add variable for the TFM build directory

The TFM build directory path is hardcoded in many places.
In order to support out-of-tree secure partitions the output path
has to be known in potentially out-of-tree build scripts.
This could potentially break out-of-tree build scripts if the
location of the build directory was changed.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

urutva commented:
@joerchan @SebastianBoe @microbuilder 

Hi, 

I've been working on creating a sample that utilises a custom secure service. This PR is used as a starting point to create the custom service. In the sample I'm creating, the custom service needs to be accessed from a file other than `main.c`. Currently, the custom service NS API defined in `main.c` and cannot be accessed from other files. It'd be better to move the custom service NS API to it's own `.c/.h` file.

Ideally, custom service NS API should be provided as part of custom service implementation and TF-M should export the header and implementation should be available in `libtfm_api_ns.a`. However, TF-M is doesn't support this for custom services.

urutva commented:
I totally agree on the organisation of the samples. But since this sample will be used as starting point for other custom services, moving NS APIs to `.c/.h` will be beneficial.

Most of the real-world implementations that would use TF-M APIs will not be implemented just in `main.c`. Therefore moving NS APIs to `.c/.h` becomes essential. If that is already done in the starting point example then it will certainly be helpful.

joerchan commented:
@urutva Thank you for your feedback, I've extracted the NS API out to its own header and source file.
Putting this in the `libtfm_api.a` library would mean building it as part of the TF-M build system. I don't think this is something that we should do.

tejlmand commented on cmake/app/boilerplate.cmake:
```diff
@@ -602,6 +602,12 @@ include(${ZEPHYR_BASE}/cmake/target_toolchain_flags.cmake)
 set(PROJECT_BINARY_DIR ${CMAKE_CURRENT_BINARY_DIR}/zephyr)
 set(PROJECT_SOURCE_DIR ${ZEPHYR_BASE})
 
+# Set TFM_BINARY_DIR. Has to be included before boards are included as some
+# boards need to define TFM build rules on the output files.
+if(CONFIG_BUILD_WITH_TFM)
+  set(TFM_BINARY_DIR ${CMAKE_BINARY_DIR}/tfm)
+endif()
```
Could you describe exactly what the problem is ?

Starting to place such specific knowledge related to a Zephyr module in boilerplate code seems wrong to me.

If someone needs to know the build folder for TF-M, why can they not use:
```
$<TARGET_PROPERTY:tfm,_EP_BINARY_DIR>
```
to obtain the path.

Or TF-M can set the property `TFM_BINARY_DIR` directly on the `tfm` target, which allows users to do:
```
$<TARGET_PROPERTY:tfm,TFM_BINARY_DIR>
```

A custom property has the advantage of being more visible to users and we can easier document it.

joerchan replied:
Thank you. That was a better suggestion. Will update to a custom target property.

joerchan replied:
@tejlmand This created the folder: `build/tfm/generated/$<TARGET_PROPERTY:tfm_cmake,TFM_BINARY_DIR>/dummy_partition/psa_manifest/tfm_dummy_partition.h`
This is coming from tfm_manifest_list.yaml.in.

joerchan replied:
I was able to fix this by using get_target_properties to store it in a variable.

tejlmand replied:
> I was able to fix this by using get_target_properties to store it in a variable.

yes, because generator expressions are not supported inside a file used by configure file, like was tried here:
https://github.com/zephyrproject-rtos/zephyr/blob/206fddecd92efe99aff1495354bdbb6161ac08fa/samples/tfm_integration/tfm_secure_partition/dummy_partition/tfm_manifest_list.yaml.in#L18

The solution you use are safe, because you do it after `find_package(Zephyr)`, so the target is known to have been created. Alternative would have been a more complex use of `add_custom_command()` to generate the needed file.

So I like this solution.

tejlmand commented on samples/tfm_integration/tfm_secure_partition/CMakeLists.txt:
```diff
@@ -0,0 +1,33 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+cmake_minimum_required(VERSION 3.13.1)
```
Please bump version:
```suggestion
cmake_minimum_required(VERSION 3.20)
```

tejlmand commented on samples/tfm_integration/tfm_secure_partition/CMakeLists.txt:
```diff
@@ -0,0 +1,34 @@
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+
+get_target_property(TFM_BINARY_DIR tfm TFM_BINARY_DIR)
+configure_file(
+  ${CMAKE_CURRENT_LIST_DIR}/dummy_partition/tfm_manifest_list.yaml.in
+  ${CMAKE_CURRENT_BINARY_DIR}/dummy_partition/tfm_manifest_list.yaml)
```
Minor nit: please decide whether you prefer:
```
configure_file(
  ${CMAKE_CURRENT_LIST_DIR}/dummy_partition/tfm_manifest_list.yaml.in
  ${CMAKE_CURRENT_BINARY_DIR}/dummy_partition/tfm_manifest_list.yaml)
```
or
```
configure_file(
  ${CMAKE_CURRENT_LIST_DIR}/dummy_partition/tfm_manifest_list.yaml.in
  ${CMAKE_CURRENT_BINARY_DIR}/dummy_partition/tfm_manifest_list.yaml
  )
```

Currently you use both styles in this file.
I'm not proposing one over the other, but I would like to see consistency inside the same file.

Please check update everywhere in this file to the style you choose.

joerchan replied:
Style chosen.
